sandbox/apparmor/notify/message.go

```diff
@@ -0,0 +1,418 @@
+	// an operation, already possessed. Use DecodeFilePermissions to decode it,
+	// if the mediation class is AA_CLASS_FILE.
+	Allow uint32
+	// Deny describes the permissions the process, attempting to perform some
+	// operation, currently lacks. Use DecodeFilePermissions to decode it, if
+	// the mediation class is AA_CLASS_FILE.
+	Deny uint32
+	// Pid of the process triggering the notification.
+	Pid uint32
+	// Label is the apparmor label of the process triggering the notification.
+	Label string
+	// Class of the mediation operation.
+	// Currently only AA_CLASS_FILE is implemented in the kernel.
+	Class MediationClass
+	// XXX: This is unused.
+	Op uint16
+}
+
+// DecodeFilePermissions returns a pair of permissions describing the state of a
+// process attempting to perform an operation.
+func (msg *MsgNotificationOp) DecodeFilePermissions() (allow, deny FilePermission, err error) {
+	if msg.Class != AA_CLASS_FILE {
+		return 0, 0, fmt.Errorf("mediation class %s does not describe file permissions", msg.Class)
+	}
+	return FilePermission(msg.Allow), FilePermission(msg.Deny), nil
+}
+
+// UnmarshalBinary unmarshals the message from binary form.
+func (msg *MsgNotificationOp) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor operation notification message"
+
+	// Unpack the base structure.
+	if err := msg.MsgNotification.UnmarshalBinary(data); err != nil {
+		return err
+	}
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	var raw msgNotificationOpKernel
+	order := arch.Endian() // ioctl messages are native byte order, verify endianness if using for other messages
+	if err := binary.Read(buf, order, &raw); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	// Unpack variable length elements.
+	unpacker := newStringUnpacker(data)
+	label, err := unpacker.UnpackString(raw.Label)
+	if err != nil {
+		return fmt.Errorf("%s: cannot unpack label: %v", prefix, err)
+	}
+
+	// Put everything together.
+	msg.Allow = raw.Allow
+	msg.Deny = raw.Deny
+	msg.Pid = raw.Pid
+	msg.Label = label
+	msg.Class = MediationClass(raw.Class)
+	msg.Op = raw.Op
+
+	return nil
+}
+
+// msgNotificationFileKernel (TBD, document me).
+//
+//	struct apparmor_notif_file {
+//	  struct apparmor_notif_op base;
+//	  uid_t suid, ouid;
+//	  __u32 name;         /* offset into data */
+//	  __u8 data[];
+//	} __attribute__((packed));
+type msgNotificationFileKernel struct {
+	msgNotificationOpKernel
+	SUID uint32
+	OUID uint32
+	Name uint32
+}
+
+// MsgNotificationFile describes a prompt to a specific file.
+type MsgNotificationFile struct {
+	MsgNotificationOp
+	SUID uint32
+	OUID uint32
+	// Name of the file being accessed.
+	// XXX: is this path valid from the point of view of the accessing process
+	// or the notify process? The name is insufficient to correctly identify
+	// the actual object being accessed in some cases.
+	Name string
```

Comment:
This is from the pov of the process being mediated. This is not currently mapped to the pov of snapd. This could be problematic, it would be better if this was in the pov of snapd but this mapping is not always possible. This is something we should discuss more.